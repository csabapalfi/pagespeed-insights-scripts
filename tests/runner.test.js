const {Runner} = require('../lib/runner');

describe('runner', () => {
  const result = {metric: 1};
  const counter =
    ({done = false, ...value} = {}) => ({next: () => ({value, done})});
  const getResult = () => Promise.resolve(result);
  const format = (x) => x;

  it('returns this as asyncIterator', () => {
    const runner = new Runner();
    expect(runner[Symbol.asyncIterator]()).toBe(runner);
  });

  it('resolves with done if counter is done', () => {
    const runner = new Runner(counter({done: true}));
    return expect(runner.next())
      .resolves.toEqual({done: true});
  });

  describe('resolves with result ', () => {
    const resolvedResult = {value: [result], done: false};
    it('that is added to samples', async () => {
      const runner = new Runner(counter(), getResult, format);
      await expect(runner.next()).resolves.toEqual(resolvedResult);
      expect(runner.samples).toContain(result);
    });

    it('that is not added to samples if warmup result ', async () => {
      const runner = new Runner(counter({warmup: true}), getResult, format);
      await expect(runner.next()).resolves.toEqual(resolvedResult);
      expect(runner.samples).not.toContain(result);
    });

    it('and no stats if last run but not enough samples (<2)', () => {
        const runner = new Runner(counter({last: true}), getResult, format);
        return expect(runner.next()).resolves.toEqual(resolvedResult);
    });

    it('and stats if last run and enough samples', () => {
        const runner = new Runner(counter({last: true}), getResult, format);
        runner.samples = [{metric: 2}];
        return expect(runner.next()).resolves.toMatchSnapshot();
    });
  });
});