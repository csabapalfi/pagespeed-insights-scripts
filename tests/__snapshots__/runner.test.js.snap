// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner resolves with result  and stats if last run and enough samples 1`] = `
Object {
  "done": false,
  "value": Array [
    Object {
      "metric": 1,
    },
    Object {
      "metric": 1.5,
      "name": "median",
      "type": "statistic",
    },
    Object {
      "metric": 0.7071067811865476,
      "name": "stddev",
      "type": "statistic",
    },
    Object {
      "metric": 1,
      "name": "min",
      "type": "statistic",
    },
    Object {
      "metric": 2,
      "name": "max",
      "type": "statistic",
    },
  ],
}
`;
